languages/csharp-dotnet-2.2/template/pipe/Module.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using Newtonsoft.Json;
+using System.Reflection;
+using __ALGO__;
+
+namespace __ALGO__.Pipe
+{
+    public class Module
+    {
+
+        private Type _primaryClassType;
+        private Type _algoInputType;
+        private MethodInfo _applyMethod;
+        
+        
+        public Module()
+        {
+            _primaryClassType = typeof(__ALGO__);
+            GetApplyMethod();
+            GetAlgoInputType();
+        }
+        
+
+        private void GetApplyMethod()
+        {
+            // getting only public methods, as those are apply methods
+            MethodInfo applyMethod = _primaryClassType.GetMethod("apply");
+            if (applyMethod == null || !applyMethod.IsStatic || !applyMethod.IsPublic)
+            {
+                throw new Exception(
+                    "No valid apply methods found. Please ensure that you're algorithm's apply function" +
+                    "is public and declared, and your primary class is named after your algorithm.");
+            }
+            _applyMethod = applyMethod;
+        }
+
+
+        private void GetAlgoInputType()
+        {
+            ParameterInfo[] parameters = _applyMethod.GetParameters();
+            if (parameters.Length == 0)
+            {
+                throw new Exception("the discovered Apply method for your Algorithm did not have any input arguments!");
+            }
+
+            ParameterInfo paramInfo = parameters[0];
+            _algoInputType = paramInfo.ParameterType;
```

Comment:
To keep things simple, we assume that the apply method has only one required input. This input can be either raw json, or a serializable class defined in __ALGO__.cs

Reply:
We also support binary, removed.